src/processor/operator/physical_operator.cpp

```diff
@@ -185,6 +185,12 @@ bool PhysicalOperator::getNextTuple(ExecutionContext* context) {
     return result;
 }
 
+void PhysicalOperator::finalize(ExecutionContext* context) {
+    for (auto& child : children) {
+        child->finalize(context);
+    }
+}
```

Comment:
1. You can add a layer of abstraction with `finalizeInternal` interface here.
2. We should skip child finalize if the operator is source.
Each operator only needs to implement the `finalizeInternal` on needs, which by default is empty (or do nothing).

```suggestion
void PhysicalOperator::finalize(ExecutionContext* context) {
    if (!isSource()) {
        for (auto& child : children) {
            child->finalize(context);
        }
    }
    finalizeInternal(context);
}
```

Reply:
updated, the one thing is I also avoid finalizing the children for `ResultCollector` since that seems to break some tests

Reply:
https://github.com/kuzudb/kuzu/pull/4067/files#r1725805914